### DISCUSSION TITLE
GitHub actions secrets protection

### DISCUSSION THREAD

**Discussion by wolflu05** [community/community#62316] [Actions] [answered]
### Select Topic Area

Question

### Body

Hello, I want to take care that my secrets can not be read by any people who propose a PR and simply change the workflow so that they send the secrets to their servers. I've read about deployment environments, codeowners, approvals, but I'm not quite sure how to do that yet correctly.

I have a workflow CI/CD which first checks the style, then runs tests and now I want to attach a job deployment which publishes a pypi package if the event is tag created. I could now just store the pipy token as secret, but how do I protect myself that no one can read it?

I would appreciate any links or examples on how to do that. I already read [this](https://docs.github.com/en/actions/security-guides/encrypted-secrets) page, but i cannot answer my question really.

**Comment by Abdullahxz** [comment]
@wolflu05 Here's an example, I hope this helps.

- Go to your Github repository in which you are creating this workflow.
- Go to Settings > Secrets and variables > Actions.
- Here you would see 2 tabs, Secrets and Variables.
- Click on secrets then select `New repository secret` button.
- Add secret name and the secret.
- After this you can access this secret in your workflow by following syntax. `${{ SECRETS.SECRET_NAME }}`

**Reply by wolflu05** [reply]
Hello @Abdullahxz thank you for your answer, but unfortunately this is not my question, I already got that far. My question is how I can prevent other guys from exposing my secrets. Imagine someone just edits the workflow so that the secrets value is send to his server instead of pypi. Then he proposed a pull request so that the workflow is triggered.

**Reply by Abdullahxz** [reply]
@wolflu05 got it, my bad. Does [this](https://blog.gitguardian.com/github-actions-security-cheat-sheet/#:~:text=Usually%2C%20we%20use%20the%20%E2%80%98pull_request%E2%80%99%20trigger%20event%20to%20trigger%20a%20workflow%20when%20someone%20submits%20a%20PR.%20With%20it%2C%20the%20triggered%20workflow%20runs%20in%20the%20context%20of%20the%20(submitter%27s)%20fork%20repo.%20Therefore%20the%20provided%20GITHUB_TOKEN%20will%20not%20have%20write%20access%20and%20the%20secrets%20are%20not%20accessible%20either.) help?

**Reply by wolflu05** [reply]
Thank you for that link, that helped a lot, I think I understood the context isolation which is there by default not correctly for the pull_request event. That makes sense.

**Comment by joshjohanning** [comment, marked as answer]
Is this an open source repository, where people will be using forks to propose changes to your workflow? 

If so, see these [docs](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/enabling-features-for-your-repository/managing-github-actions-settings-for-a-repository#controlling-changes-from-forks-to-workflows-in-public-repositories). The default setting requires approval of workflows to run for all outside collaborators, which means other people won't be able to extract your secrets.

**Reply by wolflu05** [reply]
Yes, I have an open source repo.

So I think I understood that wrong before. So is it correct now that the pull_request event (not pull_request_target) is triggered for forks that create a pull request in my repo. They also run in the actions tab of my repo, but they don't have access to my secrets because they run in their own context? So by default nobody is able to access my secrets if I also pin the action versions?

**Reply by joshjohanning** [reply]
> So is it correct now that the pull_request event (not pull_request_target) is triggered for forks that create a pull request in my repo

yes, that is how I understand it.

> They also run in the actions tab of my repo, but they don't have access to my secrets because they run in their own context? So by default nobody is able to access my secrets if I also pin the action versions?

Yes! You can also see this text when adding a secret on a repo:

> Anyone with collaborator access to this repository can use these secrets and variables for actions. **They are not passed to workflows that are triggered by a pull request from a fork**.

This can be an annoying if you WANT someone to be able to use a secret from an open source contribution, but then at that point the likely best answer is for them to be added as a collaborator to the project.

**Reply by wolflu05** [reply]
Thank you very much for the help. That clarifies it.
